Stef/conservation

stephaniesamm opened the pull request:
The files in the example_input and example_output include only the first 100 lines of my trial run for the post process pipeline with conservation, for each file. The main files that need review are the post-process pipeline (and README) and the new conservation folder.

dimostzim reviewed (approved):
conservation folder and pipeline look good to me. Also the output files have the exact format we want

katarinagresova commented on code/conservation/README.md:
```diff
@@ -0,0 +1,59 @@
+### Features:
+- Reads gene data from a TSV file or standard input
+- Processes data in blocks grouped by gene
+- Adds phyloP and phastCons scores from BigWig files
```
Are both of these files required?

stephaniesamm replied:
Phylop are evolutionary conservation at individual nucleotide sites and phastcons is probabilistic measure of conserved elements over a broader region (like a gene or regulatory sequence), so we want both.

katarinagresova commented on code/conservation/add_conservation_scores.py:
```diff
@@ -0,0 +1,120 @@
+import warnings 
+
+# Function to yield blocks of rows with the same gene from a DataFrame (alphabetically) sorted by gene sequence
+def yield_gene_blocks_from_df(df):
```
This is not efficient. You first load the whole dataset into memory and then you are iterating over it.
two options how to change it:
1. Change to the version as we have for generating negatives - yield gene block from file.
2. Load all the data into memory (into df) and just run `apply` on it. Do not use iterrows()

stephaniesamm replied:
Went with second option to avoid opening and closing bigwig files per gene block (which is sometimes just one row).

katarinagresova commented on code/conservation/add_conservation_scores.py:
```diff
@@ -0,0 +1,120 @@
+        if start >= chrom_size or end <= 0 or start >= end:
+            print(f"Invalid interval for {chrom}:{start}-{end}")
+            return [float('nan')] * (end - start)
+        start = max(0, start)
```
Shouldn't we do the checks here already - if start is <0 or end is > chrom_length, there is some problem

katarinagresova replied:
You don't need to do start = max(0, start) since you added check before

katarinagresova commented on code/conservation/add_conservation_scores.py:
```diff
@@ -0,0 +1,120 @@
+
+def add_conservation(df, phyloP_path, phastCons_path, ofile):
+    # Open the BigWig files for phyloP and phastCons
+    bw_phyloP = pyBigWig.open(phyloP_path)
```
This is dangerous. You are opening a file without using `with open...`. If the program crashes before calling `close()`, the file will stay open and you will not be able to access it again.

stephaniesamm replied:
Good point. Changed to with open.

katarinagresova commented on code/conservation/validate_conservation_scores.py:
```diff
@@ -0,0 +1,82 @@
+import pandas as pd
```
What is the reasoning of doing validation as a separate script? And not as a part of `add_conservation` script?

stephaniesamm replied:
The script were hand-me-downs from Dimos and I kept them as is (separate) at the time. 

I merged them into one now. Just adding valid conservation scores to the input file, otherwise adding NaN. 

**NOTE:** Also let's not merge this branch until the miRBench manuscript is final and submitted with a tagged version of the post-process pipeline on the main branch.

katarinagresova reviewed (commented):
Can we check that conservation is correct?
- look up few coordinates in genome browser and verify conservation scores
- look up few coordinates in genome browser that were not assigned a conservation score and see if the scores are missing in browser as well

stephaniesamm commented:
I refactored conservation code to add scores to df if they pass all validation tests, otherwise adding np.nan. 

There was also a silent problem. `float(np.nan)` returns True. I added a check with `.isnan()` in the `is_valid_number` function to catch this. 

I cannot test coordinates in genome browser as I am not sure they use the same .bw files I downloaded (however, I tested some missing scores and they are indeed missing in genome browser but the ones with the scores do not match exactly). David confirmed this might be a problem to do these checks. 

If there are no further comments I will run the pipeline to get the Manakov datasets with conservation.

stephaniesamm commented:
**REMINDER NOT TO MERGE THIS BRANCH YET.**

Let's finish up the miRBench paper first, then decide what to do with this as it is not part of the miRBench paper.

katarinagresova commented on code/post_process/example_output/example_intermediate/concat.unified_length_all_types_unique_high_confidence_deduplicated_data.tsv:
Is there a reason to have `example_intermediate/` inside of `example_output` now? Before these two folders were on the same level

stephaniesamm replied:
Just because intermediate is also an output of the pipeline, to keep parent dirs to a minimum. They can be on the same level. Moving intermediate to same level as input and output dirs.

katarinagresova commented on code/conservation/README.md:
Can we reference `scores.md` file somewhere in README? For example, if there is some term mentioned here and it is explained in `scores.md`, we can make a hyperlink to it

stephaniesamm commented on code/conservation/README.md:
```diff
@@ -0,0 +1,64 @@
+- Automatically adds 'chr' prefix if missing
+- Converts "MT" to "chrM" for mitochondrial DNA
+- Validates chromosome names against BigWig files
+- Uses 1-based input coordinates, converts to 0-based for processing
```
Yes with this script we expect 1-based input coordinates and the BigWig files have 0-based coordinates. I clarified it in the README. 

@dimostzim Do you remember the reason why you needed to do the conversion? How do we know that the HybriDetector output (and therefore input to this script) has 1-based coordinates not 0-based? I spent some time looking into this today and I don't have a sure conclusion yet.

dimostzim replied:
Yes HybridDetector output uses 1-based coordinates (same as NCBI). This is how we know:

If I open a random entry form HD for example:

gene_name: VMA21
seq.g: AACAGTAATAATAGCAATAGCTAACAACATCTGCACAGCACCTTACAGTT
start.g: 151408959
end.g: 151409008

And we go to NCBI (we know NCBI uses 1-based coordinates because it is specified in their documentation), and we search for that exact seq.g we get:

<img width="545" alt="Screenshot 2024-11-04 at 15 34 46" src="https://github.com/user-attachments/assets/53f000a2-7ea8-4356-9764-be9547eeb77e">

They are the same coordinates as strart.g and end.g  so our input uses 1-based coordinates too. the bigwig files are 0-based because the UCSC Genome Browser uses 0-based which is also specified in their documentation. 

1-based is most commonly used because NCBI uses it.

stephaniesamm commented on code/conservation/scores.md:
```diff
@@ -0,0 +1,141 @@
+   * **phastCons**: Focuses on conservation detection
+
+### Standard Thresholds
+* **phyloP**: |2| typically used as significance threshold
```
I think @dimostzim knows more about this than me.

dimostzim replied:
phyloP scores are computed as: phyloP = -log10(p-value) 
A score of +2 corresponds to a p-value of 0.01 which is a commonly used significance level.

We see such extreme cases (-20 and +20) because for example when many species (99 others) show identical sequences: probability of this happening by chance becomes extremely small

katarinagresova replied:
Can we document this a little. Or at least change `|2|` to something else? It took me some time to understand that is it absolute value (I think)

stephaniesamm commented:
Parts of this code has been added to another branch with the new post process pipeline for creating the unbiased (miRNA family balanced) datasets and merged to main. This branch is no longer valid.
